File: euler/RandomizedBinarySearch.py
def B(n):
    x = 1;
    while 2**x - 1 < n:
        x+=1

    last = n - (2**(x - 1) - 1)

    res = 0;
    for i in range(1, x):
        res += i * (2**(i - 1)) / n
    res += last * x / n

    return res

def R(n):
    res = 1
    for i in range(2, n + 1):
        ii = i * i
        res = (ii - 1) / ii * res - (ii - 2 * i + 1) / ii + 1

    return res

MM = int(1e10)
b = B(MM)
print(b)
r = R(MM)
print(b, r, r - b)

File: euler/RandomizedBinarySearch.cpp
#include <iostream>
#include <cmath>
#include <vector>
#include <iomanip>

using namespace std;

double B(long long n) {
    int x = 1;
    while (pow(2, x) - 1 < n) x++;

    int last = n - (pow(2, x - 1) - 1);

    double res = 0;
    for (int i = 1; i < x; i++) {
        res += double(i) * double(pow(2, i - 1)) / double(n);
    }
    res += double(last) * double(x) / double(n);
    return res;
}

double R(long long n) {
    double res = 1;
    for (long long i = 2; i <= n; i++) {
        res = (double(i * i) - 1) / double(i * i) * res - double(i - 1) * double(i - 1) / (i * i) + 1;
    }

    return res;
}

int main() {
    long long MM = 1e10;
    double b = B(MM);
    double r = R(MM);
    cout << std::setprecision(8) << b << " " << r << " " << r - b << endl;

    return 0;
}
